sklearn/decomposition/minibatch_nmf.py

```diff
@@ -0,0 +1,253 @@
+
+    hashing_n_features: int, default=2**10
+        Number of features for the HashingVectorizer. Only relevant if
+        hashing=True.
+
+    hashing: boolean, default=True
+        If true, the weight matrix W is rescaled at each iteration
+        to have an l1 norm equal to 1 for each row.
+
+    tol: float, default=1E-3
+        Tolerance for the convergence of the matrix W
+
+    mix_iter: int, default=2
+
+    max_iter: int, default=10
+
+    ngram_range: tuple, default=(2, 4)
+
+    init: str, default 'k-means++'
+        Initialization method of the W matrix.
+
+    random_state: default=None
+
+    Attributes
+    ----------
+
+    References
+    ----------
+    """
+
+    def __init__(self, n_components=10, batch_size=512,
+                 r=.001, hashing=False,
+                 hashing_n_features=2**12, init='k-means++',
+                 tol=1E-4, min_iter=2, max_iter=5, ngram_range=(2, 4),
+                 add_words=False, random_state=None,
+                 rescale_W=True, max_iter_e_step=20):
+
+        self.n_components = n_components
+        self.r = r
+        self.batch_size = batch_size
+        self.tol = tol
+        self.hashing = hashing
+        self.hashing_n_features = hashing_n_features
+        self.max_iter = max_iter
+        self.min_iter = min_iter
+        self.init = init
+        self.add_words = add_words
+        self.random_state = check_random_state(random_state)
+        self.rescale_W = rescale_W
+        self.max_iter_e_step = max_iter_e_step
+
+    def _rescale_W(self, W, A, B):
+        epsilon = 1E-10
+        s = W.sum(axis=1, keepdims=True)
+        s[s == 0] = epsilon
+        W /= s
+        A /= s
+        return W, A, B
+
+    def _rescale_H(self, V, H):
+        epsilon = 1e-10  # in case of a document having length=0
+        H *= np.maximum(epsilon, V.sum(axis=1).A)
+        H /= H.sum(axis=1, keepdims=True)
+        return H
+
+    def _e_step(self, Vt, W, Ht,
+                tol=1E-3, max_iter=20):
+        if self.rescale_W:
+            W_WT1 = W
+        else:
+            WT1 = np.sum(W, axis=1)
+            W_WT1 = W / WT1.reshape(-1, 1)
+        squared_tol = tol**2
+        squared_norm = 1
+        for iter in range(max_iter):
+            if squared_norm <= squared_tol:
+                break
+            Ht_W = _special_sparse_dot(Ht, W, Vt)
+            Ht_W_data = Ht_W.data
+            Vt_data = Vt.data
+            np.divide(Vt_data, Ht_W_data, out=Ht_W_data,
+                      where=(Ht_W_data != 0))
+            Ht_out = Ht * safe_sparse_dot(Ht_W, W_WT1.T)
+            squared_norm = np.linalg.norm(
+                Ht_out - Ht) / (np.linalg.norm(Ht) + 1E-10)
+            Ht[:] = Ht_out
+        return Ht
+
+    def _m_step(self, Vt, W, A, B, Ht, iter):
+        Ht_W = _special_sparse_dot(Ht, W, Vt)
+        Ht_W_data = Ht_W.data
+        np.divide(Vt.data, Ht_W_data, out=Ht_W_data, where=(Ht_W_data != 0))
+        self.rho = self.r ** (1 / (iter + 1))
```

Comment:
Any parameter that is not specified by the init, but rather a consequence of the fit should have a name that ends with "_": "self.rho" -> "self.rho_". It should also be the case of many other variables in this object.